sled-diagnostics/src/lib.rs

```diff
@@ -29,6 +31,59 @@ pub use crate::queries::{
 };
 use queries::*;
 
+/// Max number of commands to run in parallel
+const MAX_PARALLELISM: usize = 50;
+
+trait ParallelCommandExecution {
+    type Output;
+
+    /// Add a command to the set of commands to be executed.
+    async fn add_command<F>(&mut self, command: F)
+    where
+        F: std::future::Future<Output = Self::Output> + Send + 'static;
+}
+
+struct MultipleCommands<T> {
+    semaphore: Arc<Semaphore>,
+    set: JoinSet<T>,
+}
+
+impl<T: 'static> MultipleCommands<T> {
+    fn new() -> MultipleCommands<T> {
+        let semaphore = Arc::new(Semaphore::new(MAX_PARALLELISM));
+        let set = JoinSet::new();
+
+        Self { semaphore, set }
+    }
+
+    /// Wait for all commands to execute and return their output.
+    async fn join_all(self) -> Vec<T> {
+        self.set.join_all().await
+    }
+}
+
+impl<T> ParallelCommandExecution for MultipleCommands<T>
+where
+    T: Send + 'static,
+{
+    type Output = T;
+
+    async fn add_command<F>(&mut self, command: F)
+    where
+        F: std::future::Future<Output = Self::Output> + Send + 'static,
+    {
+        let permit = Arc::clone(&self.semaphore)
+            .acquire_owned()
+            .await
+            .expect("semaphore acquire");
+        let _abort_handle = self.set.spawn(async move {
```

Comment:
So, this will not even _spawn_ the tasks until there's capacity in the concurrency limit. What happens if you change it to:
```suggestion
        let semaphore = self.semaphore.clone();
        let _abort_handle = self.set.spawn(async move {
            let permit = semaphore.acquire().await.expect("semaphore acquire");
```

This way, all the tasks are spawned immediately, and the task adding commands to the set can do so synchronously (changing `add_command` to a normal `fn`) and then just wait for them to all come back. Right now, the task that adds commands has to get woken up a bunch of times to add the next one to the set; I wonder whether changing this to spawn everything synchronously would make a meaningful difference in performance...

Reply:
With this change I see the following...

stable rust + 1gb RSS:
```
took 24.416534527s
```

beta rust + 1gb RSS:
```
took 3.829577938s
```

Over chat I told you that with beta rust and the joinset we saw `took 3.674288649s` so it's negligible but overall I think a better design to go with your suggestion.  I am going to push a commit to this branch.

Reply:
ah well, it was worth a shot, i guess! thanks for giving it a spin!